### ISSUE TITLE
Android 8.0 Background Limits

### ISSUE THREAD

**Issue by thermatk** [Telegram-FOSS-Team/Telegram-FOSS#172]
Looks like always-on push service will die:
[Background Service Limitations](https://developer.android.com/about/versions/oreo/background.html#services)
Google is pushing for FCM as the only way messaging apps can function, but the [open source client](https://github.com/microg/android_external_GmsLib) doesn't look ready for production.
Let's see what upstream comes up with 😄 

Options:
- Upstream comes up with something cool
- Using microg's client library for push
- Persistent notification
- Being ok with a JobService running once in 15+ minutes

**Comment by mastad0n** [comment]
Conversations seems to have solved this problem by using a proxy service in between the user and GCM/FCM



> How do XEP-0357: Push Notifications work?

You need to be running the Play Store version of Conversations and your server needs to support push notifications.¹ Because Google Cloud Notifications (GCM) are tied with an API key to a specific app your server can not initiate the push message directly. Instead your server will send the push notification to the Conversations App server (operated by us) which then acts as a proxy and initiates the push message for you. The push message sent from our App server through GCM doesn’t contain any personal information. It is just an empty message which will wake up your device and tell Conversations to reconnect to your server. The information send from your server to our App server depends on the configuration of your server but can be limited to your account name. (In any case the Conversations App server won't redirect any information through GCM even if your server sends this information.)


https://github.com/siacs/Conversations#how-do-xep-0357-push-notifications-work

**Comment by thermatk** [comment]
@mastad0n which just reiterates the problem of not having a reliable FCM foss library

**Comment by omkarnathsingh** [comment]
Finally, Telegram is working on adapting changes for Android O. 
They added this (atleast) in yesterday's beta version. 
`<manifest xmlns:android="http://schemas.android.com/apk/res/android" android:installLocation="auto" package="org.telegram.messenger.beta" platformBuildVersionCode="26" platformBuildVersionName="8.0.0">`

**Comment by utack** [comment]
hey sorry to hijack this issue report, but I hope you guys know:  
When using Android 8.0, will the current FOSS version of Telegram not receive messages?  
Thank you!

**Comment by thermatk** [comment]
@utack restrictions should only apply for apps built with O in mind, so everything should work fine

**Comment by utack** [comment]
@thermatk appreciate your help, thank you!

**Comment by danialbehzadi** [comment]
GCM is not an option. Most of Telegram-FOSS users are using it, because they don't want anything from google  on their devices.

**Comment by Lesik** [comment]
I agree with @danialbehzadi. This time we might be able to circumvent Google's power grab with proxies. Next time they will implement restrictions that are not as easy to bypass. Remember, Google has little interest in keeping rebellious apps like ours alive. We should prioritize a non-GSM solution if possible.

**Comment by thermatk** [comment]
Upstream didn't come up with a solution yet, continuing to target sdk 25 in v4.6

I think the best approach would be to have a persistent notification in a special notification channel, so that one could hide it without affecting other notifications. Yet upstream can kill the background service in code when it's not useful for them anymore, forcing us to reimplement and maintain it 😢

**Comment by X-Ryl669** [comment]
Another option would be to flood's Google's servers with encrypted notifications. Since there is no limit to the number of messages (only to the size), that could stop them to limit such behavior, although I high doubt it. If any user of Telegram received 86400 notification per day (one per second), that might make a difference, but even so, I doubt.

**Comment by mirh** [comment]
> Upstream didn't come up with a solution yet, continuing to target sdk 25 in v4.6

[They'll like better have](https://www.xda-developers.com/play-store-updated-requirements-api-level-64-bit/) soon.

**Comment by testbird** [comment]
Maybe have a look and discuss the idea that FOSS projects could consolidate on an alternative standard to do push messages that leverages on the IMAP push:
Start at https://github.com/nextcloud/talk-android/issues/58#issuecomment-364872566

(CC @Bubu from related #216)

**Comment by iteco** [comment]
@thermatk 
you wrote for the options: 
> Being ok with a JobService running once in 15+ minutes

Why it could not be run every second ?
https://developer.android.com/reference/android/app/job/JobInfo.html#getIntervalMillis()

--
Otherways.. +1 for consolidated FOSS project for push messages.

**Comment by iteco** [comment]
Hmm... this would be a trick, but could the JobService be used to start the app's background service again?

**Comment by X-Ryl669** [comment]
No. In Doze mode, there's only few way to wake up, and the JobService is not one of them.
Only available are: GCM, Alarm (with min period of 15mn IIRC)
See [here](https://developer.android.com/training/monitoring-device-state/doze-standby.html)

**Comment by mirh** [comment]
> Only available are: GCM

I guess like it make sense to "reunite" all the *outbound* activities under a single common umbrella. 
And for as much anything is technically possible, I cannot see how any reasonable solution would involve a telegram-only-side reinvention of the wheel. 

Imo, it's [microg](https://github.com/microg/android_packages_apps_GmsCore/issues/338) or nothing.

**Comment by testbird** [comment]
> I cannot see how any reasonable solution would involve a telegram-only-side reinvention of the wheel.
> Imo, it's microg or nothing.

Or implementing https://gitlab.com/foss-push/planning, to not depend on google servers.

**Comment by mirh** [comment]
Mhh, I dunno - there doesn't seem to be lots of code (or any at all, that is)
(also, I'm not sure why not implementing that already into microg then)

**Comment by testbird** [comment]
Right, currently there is only the library from the https://delta.chat messenger, that shows that it is possible.

MicroG isn't used on other FOSS phone OSes, but yes:
https://github.com/microg/android_packages_apps_GmsCore/issues/486

**Comment by ale5000-git** [comment]
Support for putting an app into temporary whitelist when an high-priority GCM message is received: https://github.com/ale5000-git/android_packages_apps_GmsCore (latest master in my fork)
This should fix the problem with Doze.

I don't have time to compile and test it currently but if someone have time I would like some feedback.

**Comment by thermatk** [comment]
@ale5000-git which GCM library are you suggesting to use here, not to say it wouldn't solve the more general problem?

**Comment by ale5000-git** [comment]
The fix is for use of apps in conjunction with microG (apps can be compiled with microG GCM library or Google one), I don't know about other cases.

**Comment by sandy-8925** [comment]
Running a job/alarm every second is a bad idea, it will result in heavy battery drain. Any push messaging solution will require server support as well, so that's ruled out (unless it's some intermediate server like in Conversations).

**Comment by thermatk** [comment]
The time has come, it gets killed :cry:

**Comment by ildar** [comment]
Ruslan, do you mean that Telegram-FOSS doesn't work/sync with MicroG GCM
implementation anymore? Or am I getting it wrong?

**Comment by thermatk** [comment]
It never synced with GCM

**Comment by utack** [comment]
Can you just whitelist it in the "battery optimization" settings of the system?

**Comment by cRaZy-bisCuiT** [comment]
@utack  And drains massively battery for me then.

**Comment by rfc2822** [comment]
@cRaZy-bisCuiT That should not happen. Its then just excluded from "app standby" which blocks synchronization. Do you have specific measurements are is this just conjecture? We'd also be happy about such feedback on https://www.davdroid.com/forums/.

**Comment by ZJaume** [comment]
I dont' know why not tp use microg gcm library

**Comment by thermatk** [comment]
@ZJaume because it doesn't work right now, and even if it would, could stop any random day

**Comment by mikken** [comment]
Did I understand correctly that turning off Push Notification channel for Telegram and disabling system drain notification is the best way for a user to deal with this? It will leave me with normal notifications, not drain energy and not kill Telegram process.

**Comment by drrossum** [comment]
@thermatk microG GCM does work for me on android N. It even worked on early versions of android O for some months, but I couldn't get it back to work after some reinstall around 2018-04, which is why I switched back to N, and microG GCM has been working fine since then for me.

**Comment by Bubu** [comment]
@drrossum but do you know any app that actually uses the gcm **client** libraries? i.e. the thing that goes into the app, not the thing that goes in your rom.

**Comment by RocketJannis** [comment]
How does Telegram X handle it? It isnt recognized by the System as "consuming" battery and sends me push notifications.
The Google Services arent activated on my device.

**Comment by thermatk** [comment]
GCM works even if you're not logged in, just have to have gapps installed

**Comment by RocketJannis** [comment]
But they are deactivated and it also says push notifications are not guaranteed because they're not active

**Comment by ikv** [comment]
Could you please add an option to use JobService? I'm perfectly fine with receiving notifications once in a 15 minutes anyway.

**Comment by ajsb85** [comment]
FOSS users should build their own images of Android to bypass this restriction.
And continue using the own push notification service of Telegram.

**Comment by cRaZy-bisCuiT** [comment]
I dont' agree with you on this point, ajsb85. I do want FOSS and privacy, but I want to benefit from automation. You can not expect from everyone who wants to honour those principles to have their own building setup. I can not tell my grandma to have have that e.g. ;)

Someone said Google is open sourcing the FCM libraries. Any news on that end?

**Comment by cRaZy-bisCuiT** [comment]
@mirh Unfortunately MicroG is somehow on hold I'd say. The guy is pretty busy and right now it does not seem like anything will be coming soon.

**Comment by Bubu** [comment]
On January 17, 2019 3:22:56 PM GMT+01:00, cRaZy-bisCuiT <notifications@github.com> wrote:
>@mirh Unfortunately MicroG is somehow on hold I'd say. The guy is
>pretty busy and right now it does not seem like anything will be coming
>soon.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/Telegram-FOSS-Team/Telegram-FOSS/issues/172#issuecomment-455188566

See: https://chaos.social/@Bubu/101431614329523211

(N.b. https://chaos.social/@Bubu/101431600813890984 but that won't be of any use for telegram as you don't self host the server.)

**Comment by ale5000-git** [comment]
@cRaZy-bisCuiT

The development of microG is continuing, a new version was also released.

**Comment by cRaZy-bisCuiT** [comment]
> @cRaZy-bisCuiT
> 
> The development of microG is continuing, a new version was also released.

Thank you very much for your efforts, it's much appreciated! Do you accept donations?

 I did not intend to do any misconceptions about the project, it was just a snapshot of my impression of the time that I had.

**Comment by ale5000-git** [comment]
I'm only an external collaborator but you can donate to the microG project [here](https://www.bountysource.com/teams/microg).

**Comment by Mat2095** [comment]
Just out of curiosity: What solution did upstream end up with? I was using it on LineageOS 17.1 without any Google services where I did receive notifications, but often delayed by a few hours. They don't use that notification-trick, but I've seen a keep-alive-service in the settings. Is that their solution or is it unreladed?

**Comment by moradi-morteza** [comment]
i read telegram source code,they used their own push service but in android i can not recognize how they had a unlimited service in background, is ther any document or guildline?

**Comment by sandy-8925** [comment]
> i read telegram source code,they used their own push service but in android i can not recognize how they had a unlimited service in background, is ther any document or guildline?

Create foreground service with notification. May or may not need a wakelock, preferably not.

**Comment by moradi-morteza** [comment]
> > i read telegram source code,they used their own push service but in android i can not recognize how they had a unlimited service in background, is ther any document or guildline?
> 
> Create foreground service with notification. May or may not need a wakelock, preferably not.

telegram did  not use forground service they had a tgnet ndk package i think they used that to keep connection alive.
